### PR TITLE
refactor(js): generators.ts and utils.ts

### PR DESCRIPTION
**Pull request by IceHe**
## Summary

- generators.ts
    - Package all existing exports into `Generators`
    - Export default `Generators`
- utils.ts
    - Fix method `UrlSafeBase64.encode` (originaly `encodeBase64`)
    - Fix method `UrlSafeBase64.decode` (originaly `decodeBase64`)

## Testing

- Pass all existing unit tests.
- Assert `rawContent = decode(encode(rawContent))`

### CODE REVIEW

File: packages/client/src/utils.ts [hunk 1]
```diff
@@ -57,8 +57,21 @@ export const decodeToken = (token: string): IDToken => {
 
 export const nowRoundToSec = () => Math.floor(Date.now() / 1000);
 
-export const encodeBase64 = (input: string) => btoa(input);
-export const decodeBase64 = (input: string) => atob(input);
+const replaceNonUrlSafeCharactersInBase64 = (rawString: string) => {
```
Comment (simeng-li) [inline packages/client/src/utils.ts:60]: https://github.com/RGBboy/urlsafe-base64/blob/master/lib/urlsafe-base64.js looks pretty handy

File: packages/client/src/generators.ts [hunk 2]
```diff
@@ -21,18 +21,18 @@ function generateRandomString(length: number) {
  * Generates random bytes and encodes them in url safe base64.
  * @param length The length of the string (before base64url-encoding) to generate.
  */
-export const generateRandomInBase64 = (length = RANDOM_STRING_MAX_LENGTH) =>
-  encodeBase64(generateRandomString(length));
+export const generateRandomInUrlSafeBase64 = (length = RANDOM_STRING_MAX_LENGTH) =>
+  encodeInUrlSafeBase64(generateRandomString(length));
```
Comment (simeng-li) [inline packages/client/src/generators.ts:25]: just `encodeUrlSafeBase64` will do

File: packages/client/src/utils.ts (hunk 1 above)
Comment (simeng-li) [reply packages/client/src/utils.ts:60]: or we could fork the usage of this package. look pretty elegant to me 

```js
const UrlSafeBase64 = {
 encode: () => {},
 decode: () => {},
}
```

File: packages/client/src/generators.ts (hunk 2 above)
Comment (simeng-li) [reply packages/client/src/generators.ts:25]: also is there a standard for us, which method should lies under `utils` which should remain in this file? looks to me `generateRandomString` & `encodeUrlSafeBase64` share a same use case. 
- both are only used in this file
- both are general string operators

either move all to `utils` or keep in `generators` works for me . Just need to align a standard

Comment (IceHe) [reply packages/client/src/generators.ts:25]: - I prefer to put all `generate*()` methods in `generators.ts`.

    I will move `generateScope()` from `utils.ts` to `generators.ts` in another PR.

**Review by wangsijie — APPROVED** [review]

File: packages/client/src/utils.ts [hunk 3]
```diff
@@ -55,5 +55,19 @@ export const decodeToken = (token: string): IDToken => {
...
-export const decodeBase64 = (input: string) => atob(input);
+const replaceNonUrlSafeCharacters = (base64String: string) =>
+  base64String
+    .replace(/\+/g, '-') // Convert '+' to '-'
+    .replace(/\//g, '_') // Convert '/' to '_'
+    .replace(/=+$/g, ''); // Remove ending '='
+const restoreNonUrlSafeCharacters = (base64String: string) =>
+  base64String
+    .replace(/-/g, '+') // Convert '-' to '+'
+    .replace(/_/g, '/'); // Convert '_' to '/'
```
Comment (gao-sun) [inline packages/client/src/utils.ts:66]: nit: the comments is redundant, see [this doc](https://www.cs.utah.edu/~germain/PPS/Topics/commenting.html#:~:text=How%20to%20comment%20Code%3A,not%20%22self%2Ddocumenting%22.).

**Review by gao-sun — CHANGES_REQUESTED** [review]

File: packages/client/src/utils.ts (hunk 3 above)
Comment (IceHe) [reply packages/client/src/utils.ts:66]: The comments are just for who is unfamiliar with regex.
But I agree with you now, there is little need to explain regex to them.

**Review by simeng-li — APPROVED** [review]
LGTM

File: packages/client/src/utils.ts [hunk 4]
```diff
@@ -55,5 +55,19 @@ export const decodeToken = (token: string): IDToken => {
...
+
+export const UrlSafeBase64 = {
+  encode: (rawString: string) => replaceNonUrlSafeCharacters(btoa(rawString)),
+  decode: (encodedString: string) => atob(restoreNonUrlSafeCharacters(encodedString)),
+  replaceNonUrlSafeCharacters,
```
[the remark this answers is not in the document]
Comment (IceHe) [reply packages/client/src/utils.ts:66]: - `replaceNonUrlSafeCharacters ` is also used by `generateCodeChallenge` method in `utils.ts`, so I decided to export it for reuse.
- `restoreNonUrlSafeCharacters` hasn't been used, so I can remove from `UrlSafeBase64`.

TBD @simeng-li :

- Move `replaceNonUrlSafeCharacters ` method out of `UrlSafeBase64`?
- DO NOT export `restoreNonUrlSafeCharacters` method?
- More suggestions?

[the remark this answers is not in the document]
Comment (gao-sun) [reply packages/client/src/utils.ts:66]: we can move them to our [essentials](https://github.com/silverhand-io/essentials) package.

[the remark this answers is not in the document]
Comment (gao-sun) [reply packages/client/src/utils.ts:66]: let's add a linear issue before merging.

**Review by gao-sun — APPROVED** [review]

File: packages/client/src/utils.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (IceHe) [reply packages/client/src/utils.ts:66]: done.

**Pull request merged**